### PR TITLE
Fix media library with PDF

### LINKED ISSUE #1071: No media displaying if PDF in WP media library
## Bug Description

No media displaying if PDF in WP media library

## Expected Behaviour

Media is displayed

## Steps to Reproduce

- Go to media in left hard sidebar
- Upload a PDF.
- Create new story.

## Screenshots
Screenshot from 2020-04-06 16-10-52

### PR DESCRIPTION
**Pull request by merapi**
Fixes #1071

Returning `null` is ok with all usages of this function.

### CODE REVIEW

**Comment by swissspidy** [comment]
Sounds like a great thing to add tests for. Not just `getTypeFromMime`, but also all the places where it's being used - especially the media library pane.

Also, shouldn't we just avoid returning PDFs in the REST API in the first place? Might save some bandwidth. We should discuss this properly in #1071 while this here can serve as a workaround.

File: assets/src/edit-story/app/media/utils/getTypeFromMime.js [hunk 1]
```diff
@@ -18,16 +18,15 @@
  * Infer element type from mime type of its resource
  *
  * @param {string} mimeType Mime type.
- * @return {string} Element type.
+ * @return {string|null} Element type.
  */
 const getTypeFromMime = (mimeType) => {
```
Comment (spacedmonkey) [inline assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: I personally have never liked this function. I think it is very limited. A mime type is `image/gif`. Why not split and shift. 
```
const getTypeFromMime = (mimeType) => mimeType.split('/').shift();
```

**Comment by spacedmonkey** [comment]
> Sounds like a great thing to add tests for. Not just `getTypeFromMime`, but also all the places where it's being used - especially the media library pane.
> 
> Also, shouldn't we just avoid returning PDFs in the REST API in the first place? Might save some bandwidth. We should discuss this properly in #1071 while this here can serve as a workaround.

So I looked into not returning unsupported file types  in the rest api. We would have to extend the existing REST API to do this. Also mime type searches in WP are pretty extensive, as they are like search. See the original trac [ticket](https://core.trac.wordpress.org/ticket/45354).

File: assets/src/edit-story/app/media/utils/getTypeFromMime.js (hunk 1 above)
Comment (merapi) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: Yep, seems reasonable, I guess this function is just for returning type, not type from our supported subset? @obetomuniz

**Comment by spacedmonkey** [comment]
The REST API only support one mime type or type are a time. https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php#L616-L629

File: assets/src/edit-story/app/media/utils/getTypeFromMime.js (hunk 1 above)
Comment (obetomuniz) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: Yup. That’s correct @merapi. It’s for normalization. PDF for instance will be `application/pdf`, returning "application" value on the split/shift method. Is it expected across the editor? What if I send a JSON file, which have the mimetype "application/json"?

Comment (merapi) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: Right, element type `application` doesn't make sense for PDF and JSON (although, not sure if we have any use case). For our current use-case - both will do the job.

Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: The type should be correct. Returning null doesn't make a lot of sense.

Comment (merapi) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: What is the definition of _correct_ element type here? `application` for PDF (or anything else) doesn't sound like one.

Comment (spacedmonkey) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: PDF is a weird example, but a type is a wide group of what the mime type is. So `image/gif` is an image, `video/mp4` is a video, `text/html` is text and `audio/mp3` is audio. So I thin the logic I provided should for all those use cases. What you do with the type is a nothing matter.

Comment (merapi) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: Just saying: for now - it does the job.
If more cases will arrive, like handling `text/html` then `application/pdf` is also likely, for example, you want to attach some kind of document (text or pdf) - we will have to change both of these functions. No such thing for now.

Comment (swissspidy) [reply assets/src/edit-story/app/media/utils/getTypeFromMime.js:23]: That works for me, but it should be documented. Ideally with some tests & inline docs.

At the very least, we should have a test verifying that the media library renders fine when there are PDFs (if possible).

**Review by swissspidy — APPROVED** [review]
LGTM but in need of tests to avoid regressions.

File: assets/src/edit-story/components/library/test/mediaPane.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * Internal dependencies
+ */
+import { arrange, mediaResponseWithPdfAndImage } from './_utils';
+
+describe('mediaPane', () => {
+  it('should only display supported mimeTypes and not crash on unsupported', async () => {
+    const { container } = await arrange({
+      mediaResponse: mediaResponseWithPdfAndImage,
+    });
+    expect(container.querySelector(`img[src*=testImage]`)).not.toBeNull();
```
Comment (barklund) [inline assets/src/edit-story/components/library/test/mediaPane.js]: Could you change this to `expect(getByAltText('test-alt')).toBeDefined()`? That's a better test pattern.

File: assets/src/edit-story/components/library/test/_utils.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  // Another option without allPromises:
+  // const flushPromises = () => new Promise(window.setImmediate);
+  // await act(flushPromises);
+
+  await act(() => Promise.all(allPromises));
+
+  return accessors;
+}
+
+export const mediaResponseWithPdfAndImage = [
```
Comment (barklund) [inline assets/src/edit-story/components/library/test/_utils.js]: Please move this huge object to an external file. Our test suite ignores `_utils` files. Or you could create a `"fixture.json"` file, which would similarly be ignored.

File: assets/src/edit-story/components/library/test/_utils.js [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        </APIContext.Provider>
+      </ConfigProvider>
+    </ThemeProvider>
+  );
+
+  // Another option without allPromises:
+  // const flushPromises = () => new Promise(window.setImmediate);
+  // await act(flushPromises);
+
+  await act(() => Promise.all(allPromises));
```
Comment (barklund) [inline assets/src/edit-story/components/library/test/_utils.js]: I'm not sure I like this being in the arrange function. But given this is a single-use arrange, it's probably okay for now.

**Review by barklund — CHANGES_REQUESTED** [review]
A few comments on the test suite.

File: assets/src/edit-story/components/library/test/mediaPane.js (hunk 2 above)
Comment (merapi) [reply assets/src/edit-story/components/library/test/mediaPane.js]: Right, alt text fix is not yet merged, so I added a small fix here.

File: assets/src/edit-story/components/library/test/_utils.js (hunk 4 above)
Comment (merapi) [reply assets/src/edit-story/components/library/test/_utils.js]: Yep, not proud of it, we can reuse this arrange for other panels, so then we will make it pretty/reusable ✨.

**Review by barklund — APPROVED** [review]
Works for me, but please have someone else with more knowledge of the media management review it too.

**Review by spacedmonkey — APPROVED** [review]

**Comment by spacedmonkey** [comment]
Travis build still failing

**Review by wassgha — APPROVED** [review]

**Pull request merged**